Import local node

michallowasrzechonek-silvair opened the pull request.

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -1366,17 +1370,386 @@ static bool get_app_properties(struct mesh_node *node, const char *path,
+	int tmp;
+	json_object *jvalue;
+
+	if (!json_object_object_get_ex(jobj, "IVindex", &jvalue))
```
see `mesh-db.c` function `get_int` - it might be a good idea to extract this for reuse

jakubwitowski-silvair replied:
Reading json will be simplified in the near future

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -1366,17 +1370,386 @@ static bool get_app_properties(struct mesh_node *node, const char *path,
+	return true;
+}
+
+static bool parse_imported_device_key(json_object *jobj, uint8_t key_buf[16])
```
same, `mesh_db_read_device_key`/`mesh_db_read_token` - all of these functions read a hexstring from json

jakubwitowski-silvair replied:
Reading json will be simplified in the near future

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -1366,17 +1370,386 @@ static bool get_app_properties(struct mesh_node *node, const char *path,
+
+		switch (strlen(jidx)) {
+
+		case 4: {
```
I don't think you need braces here

jakubwitowski-silvair replied:
Removed

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -1366,17 +1370,386 @@ static bool get_app_properties(struct mesh_node *node, const char *path,
+		return false;
+	}
+
+	if (node_uuid_get(node))
```
I don't think this `if` is needed, node objects always have uuids set.

jakubwitowski-silvair replied:
Removed

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -1672,13 +2040,108 @@ void node_join(const char *app_path, const char *sender, const uint8_t *uuid,
+	struct mesh_node *node;
+	struct mesh_net *net;
+	struct mesh_db_node db_node;
+	struct keyring_net_key *net_keyring;
```
Unused variable?

jakubwitowski-silvair replied:
Correct

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -1561,29 +1922,36 @@ static void get_managed_objects_cb(struct l_dbus_message *msg, void *user_data)
-						DEFAULT_IV_INDEX, dev_key,
-						PRIMARY_NET_IDX,
-						net_key.old_key))
+		l_getrandom(dev_key, 16);
```
I would leave the `sizeof` and use that in other places too, instead of hardcoding `16`

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -1366,17 +1370,386 @@ static bool get_app_properties(struct mesh_node *node, const char *path,
+	iend = json_object_iter_end(jobject);
+
+	while (!json_object_iter_equal(&iter, &iend)) {
+		net_keyring = l_new(struct keyring_net_key, 1);
```
`mesh/node.c:1483:3: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]`

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -1366,17 +1370,386 @@ static bool get_app_properties(struct mesh_node *node, const char *path,
+
+static void add_keys(void *net_keys, void *_node)
+{
+	l_info("ADD KEYS CB");
```
`mesh/node.c:1725:2: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]`

jakubwitowski-silvair replied:
log removed

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -1550,7 +1911,7 @@ static void get_managed_objects_cb(struct l_dbus_message *msg, void *user_data)
 		node_ready_func_t cb = req->cb;
-		struct keyring_net_key net_key;
+		net_key = l_new(struct keyring_net_key, 1);
 		uint8_t dev_key[16];
```
`mesh/node.c:1915:3: error: ISO C90 forbids mixed declarations and code [-Werror=declaration-after-statement]`

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -1672,13 +2040,108 @@ void node_join(const char *app_path, const char *sender, const uint8_t *uuid,
 					node_ready_func_t cb, void *user_data)
 {
 	struct managed_obj_request *req;
+	struct mesh_node *node;
```
`mesh/node.c:2140:20: error: variable ‘node’ set but not used [-Werror=unused-but-set-variable]`

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -1426,6 +1785,8 @@ static void get_managed_objects_cb(struct l_dbus_message *msg, void *user_data)
 	const char *path;
 	struct mesh_node *node = NULL;
+	struct l_queue *net_keys;
+	struct keyring_net_key *net_key;
```
`mesh/node.c:1953:3: error: ‘net_key’ may be used uninitialized in this function [-Werror=maybe-uninitialized]`, initialize to `NULL`

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -1426,6 +1785,8 @@ static void get_managed_objects_cb(struct l_dbus_message *msg, void *user_data)
 	struct managed_obj_request *req = user_data;
 	const char *path;
 	struct mesh_node *node = NULL;
+	struct l_queue *net_keys;
```
`mesh/node.c:1949:5: error: ‘net_keys’ may be used uninitialized in this function [-Werror=maybe-uninitialized]`

michallowasrzechonek-silvair commented on doc/mesh-api.txt:
```diff
@@ -151,7 +151,7 @@ Methods:
 			org.bluez.mesh.Error.AlreadyExists,
 
-	 uint64 token ImportLocalNode(string json_data)
+	 uint64 token ImportLocalNode(string json_data, array{byte}[16] uuid)
```
Update API documentation

michallowasrzechonek-silvair replied:
Also, add documentation about `json_data` schema

michallowasrzechonek-silvair commented on mesh/mesh.c:
```diff
@@ -758,12 +759,56 @@ static struct l_dbus_message *create_network_call(struct l_dbus *dbus,
 	l_queue_push_tail(pending_queue, pending_msg);
 
-	node_create(app_path, sender, uuid, create_network_ready_cb,
+	node_create(sender, app_path, uuid, create_node_ready_cb,
```
Why change the order of arguments?

michallowasrzechonek-silvair commented on mesh/mesh.c:
```diff
@@ -758,12 +759,56 @@ static struct l_dbus_message *create_network_call(struct l_dbus *dbus,
+	jnode = json_tokener_parse(json_data);
+
+	sender = l_dbus_message_get_sender(msg);
+	pending_msg = l_dbus_message_ref(msg);
```
If `node_import` fails, there is noone to `unref` the message, and you leak memory.

michallowasrzechonek-silvair commented on mesh/model.c:
```diff
@@ -228,6 +228,7 @@ static struct l_dbus_message *create_config_update_msg(struct mesh_node *node,
 
 	if ((id & VENDOR_ID_MASK) != VENDOR_ID_MASK) {
 		uint16_t vendor = id >> 16;
+
```
No need for blank lines here, you can leave the whole file unmodified.

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -27,6 +27,9 @@
 #include <json-c/json.h>
+#include <stdio.h>
+#include <string.h>
+#include <assert.h>
```
Leftover includes

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -61,6 +64,7 @@
 #define REQUEST_TYPE_JOIN 0
 #define REQUEST_TYPE_ATTACH 1
 #define REQUEST_TYPE_CREATE 2
+#define REQUEST_TYPE_IMPORT 3
```
Maybe make this an enum?

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -848,136 +863,6 @@ uint16_t node_generate_comp(struct mesh_node *node, uint8_t *buf, uint16_t sz)
 }
 
 
-#define MIN_COMPOSITION_LEN 16
```
This belongs to a separate patch

michallowasrzechonek-silvair commented on mesh/node.h:
```diff
@@ -89,8 +89,10 @@ int node_attach(const char *app_path, const char *sender, uint64_t token,
-void node_create(const char *app_path, const char *sender, const uint8_t *uuid,
+void node_create(const char *sender, const char *app_path, const uint8_t *uuid,
 					node_ready_func_t cb, void *user_data);
+bool node_import(const char *app_path, const char *sender, void *jnode,
```
Please keep arguments in the same order.

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -1547,11 +1507,49 @@ static void get_managed_objects_cb(struct l_dbus_message *msg, void *user_data)
+
 	} else {
 		/* Callback for create node request */
+		uint8_t dev_key[16];
```
Please don't reorder lines if you don't touch the logic. As a rule of thumb, try to keep the patch as small as possible.

michallowasrzechonek-silvair commented on doc/mesh-api.txt:
```diff
@@ -151,16 +151,34 @@ Methods:
+		Elements) it is validated against composition data provided by
+		the application. Otherwise, new node is created based on
+		composition data provided by the application using provisioning data
+		from import_data parameter.
```
Mention that sequenceNumber is optional

michallowasrzechonek-silvair commented on doc/mesh-api.txt:
```diff
@@ -151,16 +151,34 @@ Methods:
+			- Simplified representation of node configuration with
+			  provisioning data only
+			- FULL representation of node configuration with both
+			  provisioning and composition data
```
"FULL" -> "Full"

michallowasrzechonek-silvair commented on doc/mesh-api.txt:
```diff
@@ -1061,6 +1079,68 @@ Properties:
+================================
+Example of Json format for ImportLocalNode():
+
+	Import local node operation:
```
Keep it consistent: "Simplified" and "Full" representation, docs for the API don't mention any "migration"

michallowasrzechonek-silvair commented on mesh/mesh-db.h:
```diff
@@ -16,6 +16,12 @@
+#define mesh_db_import_node_foreach(obj, key, val) \
+				json_object_object_foreach(obj, key, val)
+
+typedef json_object parsed_type;
```
Don't typedef.

michallowasrzechonek-silvair commented on mesh/mesh-db.h:
```diff
@@ -93,6 +99,10 @@ struct mesh_db_prov {
 };
 
+struct mesh_db_import_node {
+	parsed_type *parsed_node;
```
That's a big :-1:. The whole idea smells - you should rather create two separate `mesh_db_import_data` structs, one for simplified flavour and the other for full import, than pass this between `mesh`, `node` and `mesh-db` modules.

Perhaps it would be even possible to reuse `mesh_db_node` for that purpose ("inherit" from it, maybe?).

michallowasrzechonek-silvair replied:
This way you wouldn't need to expose that much details in `mesh-db.h`

michallowasrzechonek-silvair commented on mesh/mesh-db.c:
```diff
@@ -253,6 +253,134 @@ static json_object *jarray_key_del(json_object *jarray, int16_t idx)
+static bool json_iter(json_object *jobj, struct mesh_db_import_node *i_node)
+{
+	bool dev_k = false, unicast = false;
+	bool ivi = false, ivu = false, net_k = false;
```
Please expand names. For example `ivi` is a different entity than `iv_index`, this is confusing!

michallowasrzechonek-silvair commented on mesh/mesh-db.c:
```diff
@@ -253,6 +253,134 @@ static json_object *jarray_key_del(json_object *jarray, int16_t idx)
+		}
+	}
+
+	if (dev_k && unicast && net_k && ivi && ivu)
```
`return dev_k && unicast && net_k && ivi && ivu`

michallowasrzechonek-silvair commented on mesh/mesh-db.c:
```diff
@@ -1783,3 +1912,34 @@ bool mesh_db_model_sub_del_all(json_object *jnode, uint16_t addr,
+struct mesh_db_import_node *mesh_db_parse_import_data(const char *import_data)
+{
+	json_object *jobj = json_tokener_parse(import_data);
+	struct mesh_db_import_node *i_node =
```
Rename to `import_node`

michallowasrzechonek-silvair commented on mesh/mesh-db.c:
```diff
@@ -1783,3 +1912,34 @@ bool mesh_db_model_sub_del_all(json_object *jnode, uint16_t addr,
+
+}
+
+bool mesh_db_import_node_is_simplified(struct mesh_db_import_node *i_node)
```
Remove this function, check for `import_node->db_node->elements` directly

michallowasrzechonek-silvair commented on mesh/mesh-db.h:
```diff
@@ -93,6 +94,16 @@ struct mesh_db_prov {
 	uint8_t priv_key[32];
 };
 
+struct mesh_db_import_node {
```
Rename to `mesh_db_import`

michallowasrzechonek-silvair commented on mesh/mesh-db.h:
```diff
@@ -93,6 +94,16 @@ struct mesh_db_prov {
 
+struct mesh_db_import_node {
+
+	struct mesh_db_node *node;
```
Rename to `db_node`

michallowasrzechonek-silvair commented on mesh/mesh-db.h:
```diff
@@ -93,6 +94,16 @@ struct mesh_db_prov {
+	uint8_t dev_key[16];
+	uint8_t net_key[16];
+	uint16_t net_idx;
+	uint32_t iv_idx;
```
Rename to 'iv_index'

michallowasrzechonek-silvair commented on mesh/mesh.c:
```diff
@@ -70,6 +72,10 @@ struct join_data{
 
+static const char * const supported_import_data_types[] = {
+	"json"
+};
```
:+1:

michallowasrzechonek-silvair commented on mesh/mesh.c:
```diff
@@ -383,6 +389,18 @@ static void node_init_cb(struct mesh_node *node, struct mesh_agent *agent)
 
+static bool validate_data_type(const char *data_type)
+{
+	uint8_t len = sizeof(supported_import_data_types) /
```
`L_ARRAY_SIZE`

michallowasrzechonek-silvair commented on mesh/mesh.c:
```diff
@@ -383,6 +389,18 @@ static void node_init_cb(struct mesh_node *node, struct mesh_agent *agent)
+	uint8_t len = sizeof(supported_import_data_types) /
+					sizeof(supported_import_data_types[0]);
+
+	for (uint8_t idx = 0; idx < len; idx++) {
```
Move declaration outside of `for`

michallowasrzechonek-silvair commented on mesh/mesh.c:
```diff
@@ -593,12 +611,73 @@ static struct l_dbus_message *create_network_call(struct l_dbus *dbus,
+
+	l_info("PARSE IMPORTED DATA");
+	p_node = mesh_db_parse_import_data(import_data);
+	l_info("PARSE IMPORTED DATA FINISHED");
```
Remove logs.

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -110,7 +112,13 @@ struct managed_obj_request {
+
+struct node_import_request {
+	uint8_t uuid[16];
+	struct mesh_db_import_node *i_node;
```
rename `mesh_db_import_node` to `mesh_db_import` and `i_node` to `import`

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -870,7 +878,7 @@ uint16_t node_generate_comp(struct mesh_node *node, uint8_t *buf, uint16_t sz)
 
 bool node_parse_composition(struct mesh_node *node, uint8_t *data,
-								uint16_t len)
+			    uint16_t len)
```
Ident?

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -965,7 +973,7 @@ bool node_parse_composition(struct mesh_node *node, uint8_t *data,
 			mod_id |= (vendor_id << 16);
 			mod = mesh_model_vendor_new(ele->idx, vendor_id,
-									mod_id);
+						    mod_id);
```
Indent?

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -996,7 +1004,6 @@ bool node_parse_composition(struct mesh_node *node, uint8_t *data,
 
 	return false;
 }
-
```
Blank line?

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -1566,6 +1565,48 @@ static void get_managed_objects_cb(struct l_dbus_message *msg, void *user_data)
+	} else if (req->type == REQUEST_TYPE_IMPORT) {
+
+		node_ready_func_t cb = req->cb;
+		struct node_import_request *import_data = req->user_data;
```
Extract `mesh_db_import` to a local variable, call to `add_local_node` will be shorter.

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -1614,14 +1655,22 @@ static void get_managed_objects_cb(struct l_dbus_message *msg, void *user_data)
+								req->user_data;
+			node_ready_func_t cb = req->cb;
+
+			cb(import_data->user_data, MESH_ERROR_FAILED, NULL);
```
Other request types call the callback with `req->user_data`, please keep this consistent here.

jakubwitowski-silvair replied:
Won't be done :)

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -1686,6 +1735,50 @@ void node_join(const char *app_path, const char *sender, const uint8_t *uuid,
 }
 
+static bool node_import_full(const char *app_path, const char *sender,
+					void *parsed_node, const uint8_t *uuid,
```
Rename `parsed_node` to `import` and use real type instead of `void *`.

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -1686,6 +1735,50 @@ void node_join(const char *app_path, const char *sender, const uint8_t *uuid,
+}
+
+static bool node_import_simplified(const char *app_path, const char *sender,
+					void *parsed_node, const uint8_t *uuid,
```
Rename `parsed_node` to `import` and use real type instead of `void *`.

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -1686,6 +1735,50 @@ void node_join(const char *app_path, const char *sender, const uint8_t *uuid,
+
+bool node_import(const char *app_path, const char *sender, void *parsed_node,
+		const uint8_t *uuid, node_ready_func_t cb, void *user_data)
+{
```
Rename `parsed_node` to `import` and cast to real type before passing to `_simplified`/`_full`

michallowasrzechonek-silvair commented:
Way better :)

michallowasrzechonek-silvair commented on mesh/mesh-db.c:
```diff
@@ -1783,3 +1909,31 @@ bool mesh_db_model_sub_del_all(json_object *jnode, uint16_t addr,
+	import_node->node = l_new(struct mesh_db_node, 1);
+
+	if (!json_iter(jobj, import_node)) {
+		json_object_put(jobj);
```
IMO lepiej byłoby zrobić `goto error` i tam wyczyścić

michallowasrzechonek-silvair replied:
Tak samo przy `if (!jobj)`

michallowasrzechonek-silvair commented on mesh/mesh.c:
```diff
@@ -593,12 +611,72 @@ static struct l_dbus_message *create_network_call(struct l_dbus *dbus,
+	if (!node_import(app_path, sender, p_node, uuid, create_node_ready_cb,
+								pending_msg)) {
+		l_free(p_node);
+		l_dbus_message_unref(msg);
```
Robisz `unref`, a na kolejce `pending_queue` zostaje dangling pointer.

michallowasrzechonek-silvair replied:
`goto`!

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -26,7 +26,7 @@
 
 #include <ell/ell.h>
-#include <json-c/json.h>
+#include <stdio.h>
```
Remove

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -110,7 +112,13 @@ struct managed_obj_request {
+
+struct node_import_request {
+	uint8_t uuid[16];
+	struct mesh_db_import *i_node;
```
Zmień nazwę pola na `import`

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -1566,6 +1565,49 @@ static void get_managed_objects_cb(struct l_dbus_message *msg, void *user_data)
+
+		node_ready_func_t cb = req->cb;
+		struct node_import_request *import_data = req->user_data;
+		struct mesh_db_import *db_node = import_data->i_node;
```
Nazwy!

```
struct node_import_request *import_request = req->user_data;
struct mesh_db_import *import = import_request->import;
```

michallowasrzechonek-silvair commented on mesh/node.c:
```diff
@@ -1686,6 +1736,53 @@ void node_join(const char *app_path, const char *sender, const uint8_t *uuid,
 }
 
+static bool node_import_full(const char *app_path, const char *sender,
+					struct mesh_db_import *import_node,
```
Samo `import` a nie `import_node` ;P

jakubwitowski-silvair committed:
mesh: Extract read_* functions in mesh-config-json

This is a small improvement of read_node function readability.

michallowasrzechonek-silvair committed:
mesh: Add documentation for Import() D-Bus API

This replaces previously specified ImportLocalNode() call

michallowasrzechonek-silvair committed:
mesh: Implement Import() D-Bus API of org.bluez.mesh.Network1 interface

This method allows local nodes to be imported from an external
provisioning database, enabling applications to create mesh nodes
without using provisioning procedure.

The procedure is similar to provisioning procedure, but all data
exchange happens over D-Bus.
